package/gluon-mesh-babel/src/respondd.c

```diff
@@ -113,8 +354,98 @@ static void obtain_if_addr(const char *ifname, char *lladdr) {
 	freeifaddrs(ifaddr);
 }
 
+static void add_neighbour(char **data) {
+	struct json_object *neigh = json_object_new_object();
+
+	if (data[RXCOST])
+		json_object_object_add(neigh, "rxcost", json_object_new_int(atoi(data[RXCOST])));
+	if (data[TXCOST])
+		json_object_object_add(neigh, "txcost", json_object_new_int(atoi(data[TXCOST])));
+	if (data[COST])
+		json_object_object_add(neigh, "cost", json_object_new_int(atoi(data[COST])));
+	if (data[REACH])
+		json_object_object_add(neigh, "reachability", json_object_new_double(strtod(data[REACH], NULL)));
+
+	struct json_object *nif = NULL;
+
+	if (data[IF] && !json_object_object_get_ex(neighbours, data[IF], &nif)) {
+		char str_ip[NI_MAXHOST] = {};
+		obtain_if_addr((const char *)data[IF], str_ip);
+
+		nif = json_object_new_object();
+		if (nif) {
+			json_object_object_add(nif, "ll-addr", json_object_new_string(str_ip));
+			json_object_object_add(nif, "protocol", json_object_new_string("babel"));
+			json_object_object_add(neighbours, data[IF], nif);
+		}
+	}
+
+	struct json_object *nif_neighbours = 0;
+	if (!json_object_object_get_ex(nif, "neighbours", &nif_neighbours)) {
+		nif_neighbours = json_object_new_object();
+		json_object_object_add(nif, "neighbours", nif_neighbours);
+	}
+
+	json_object_object_add(nif_neighbours, data[ADDRESS], neigh);
+}
+
+static void del_neighbour(char **data) {
+	struct json_object *nif;
+	if (json_object_object_get_ex(neighbours, data[IF], &nif)) {
+		struct json_object *neighbour = 0;
+		if (json_object_object_get_ex(nif, "neighbours", &neighbour)) {
+			json_object_object_del(neighbour, data[ADDRESS]);
+		}
+	}
+}
+
+static bool handle_neighbour(char **data, void *obj) {
+	pthread_mutex_lock(&neighbours_lock);
```

Comment:
The mutex calls should be inside the `if` block to reduce unnecessary empty `lock`-`unlock` pairs.

Reply:
(y)